File: bucket_3/5.ts
import * as grpc from 'grpc';
import { Empty } from 'google-protobuf/google/protobuf/empty_pb';
import { AccountServiceService, ChildrenFollowingServiceService, TweetsServiceService } from './generated/twitter_grpc_pb';
import { Child, Parent, Tweet, TweetComment } from './generated/twitter_pb';

class AccountService implements AccountServiceService {
  createParent(call: grpc.ServerUnaryCall<Parent>, callback: grpc.sendUnaryData<Parent>): void {
    // Implement logic to create a parent account
    const parent = new Parent();
    callback(null, parent);
  }

  updateParent(call: grpc.ServerUnaryCall<Parent>, callback: grpc.sendUnaryData<Parent>): void {
    // Implement logic to update a parent account
    const parent = new Parent();
    callback(null, parent);
  }

  createChild(call: grpc.ServerUnaryCall<Child>, callback: grpc.sendUnaryData<Child>): void {
    // Implement logic to create a child account
    const child = new Child();
    callback(null, child);
  }

  updateChild(call: grpc.ServerUnaryCall<Child>, callback: grpc.sendUnaryData<Child>): void {
    // Implement logic to update a child account
    const child = new Child();
    callback(null, child);
  }

  getChildByUsername(call: grpc.ServerUnaryCall<{ username: string }>, callback: grpc.sendUnaryData<Child>): void {
    // Implement logic to get a child by username
    const child = new Child();
    callback(null, child);
  }

  verifyEmail(call: grpc.ServerUnaryCall<{ key: string }>, callback: grpc.sendUnaryData<Empty>): void {
    // Implement logic to verify email
    const response = new Empty();
    callback(null, response);
  }
}

class ChildrenFollowingService implements ChildrenFollowingServiceService {
  followChild(call: grpc.ServerUnaryCall<{ childId: string }>, callback: grpc.sendUnaryData<Empty>): void {
    // Implement logic to follow a child
    const response = new Empty();
    callback(null, response);
  }

  listFollowers(call: grpc.ServerUnaryCall<{ childId: string }>): void {
    // Implement logic to list followers of a child
    const followers: Child[] = [];
    const stream = call.request.childId;
    for (const follower of followers) {
      stream.write(follower);
    }
    stream.end();
  }

  listFollowing(call: grpc.ServerUnaryCall<{ childId: string }>): void {
    // Implement logic to list following of a child
    const following: Child[] = [];
    const stream = call.request.childId;
    for (const followed of following) {
      stream.write(followed);
    }
    stream.end();
  }
}

class TweetsService implements TweetsServiceService {
  createTweet(call: grpc.ServerUnaryCall<Tweet>, callback: grpc.sendUnaryData<Tweet>): void {
    // Implement logic to create a tweet
    const tweet = new Tweet();
    callback(null, tweet);
  }

  getTweet(call: grpc.ServerUnaryCall<{ id: string }>, callback: grpc.sendUnaryData<Tweet>): void {
    // Implement logic to get a tweet
    const tweet = new Tweet();
    callback(null, tweet);
  }

  deleteTweet(call: grpc.ServerUnaryCall<{ id: string }>, callback: grpc.sendUnaryData<Empty>): void {
    // Implement logic to delete a tweet
    const response = new Empty();
    callback(null, response);
  }

  likeTweet(call: grpc.ServerUnaryCall<{ id: string }>, callback: grpc.sendUnaryData<Empty>): void {
    // Implement logic to like a tweet
    const response = new Empty();
    callback(null, response);
  }

  listTweets(call: grpc.ServerUnaryCall<{ childId: string }>): void {
    // Implement logic to list tweets
    const tweets: Tweet[] = [];
    const stream = call.request.childId;
    for (const tweet of tweets) {
      stream.write(tweet);
    }
    stream.end();
  }

  listFeedTweets(call: grpc.ServerUnaryCall<Empty>): void {
    // Implement logic to list feed tweets
    const feedTweets: Tweet[] = [];
    const stream = call.request;
    for (const tweet of feedTweets) {
      stream.write(tweet);
    }
    stream.end();
  }

  createTweetComment(call: grpc.ServerUnaryCall<TweetComment>, callback: grpc.sendUnaryData<TweetComment>): void {
    // Implement logic to create a tweet comment
    const tweetComment = new TweetComment();
    callback(null, tweetComment);
  }

  getTweetComment(call: grpc.ServerUnaryCall<{ id: string }>, callback: grpc.sendUnaryData<TweetComment>): void {
    // Implement logic to get a tweet comment
    const tweetComment = new TweetComment();
    callback(null, tweetComment);
  }
}

function startServer(): void {
  const server = new grpc.Server();
  server.addService(AccountServiceService, new AccountService());
  server.addService(ChildrenFollowingServiceService, new ChildrenFollowingService());
  server.addService(TweetsServiceService, new TweetsService());
  server.bind('0.0.0.0:50051', grpc.ServerCredentials.createInsecure());
  server.start();
}

startServer();


// client:

// Get a child by username
function getChildByUsername(username: string) {
  const request = new GetChildByUsernameRequest();
  request.setUsername(username);

  client.getChildByUsername(request, (error: grpc.ServiceError | null, response: Child) => {
    if (!error) {
      console.log('Child details:', response.toObject());
    } else {
      console.error('Error getting child by username:', error.details);
    }
  });
}

File: bucket_3/5.py
import grpc
from concurrent import futures
from google.protobuf import empty_pb2
from generated import twitter_pb2, twitter_pb2_grpc

class AccountService(twitter_pb2_grpc.AccountServicer):
    def CreateParent(self, request, context):
        # Implement logic to create a parent account
        parent = twitter_pb2.Parent()
        return parent

    def UpdateParent(self, request, context):
        # Implement logic to update a parent account
        parent = twitter_pb2.Parent()
        return parent

    def CreateChild(self, request, context):
        # Implement logic to create a child account
        child = twitter_pb2.Child()
        return child

    def UpdateChild(self, request, context):
        # Implement logic to update a child account
        child = twitter_pb2.Child()
        return child

    def GetChildByUsername(self, request, context):
        # Implement logic to get a child by username
        child = twitter_pb2.Child()
        return child

    def VerifyEmail(self, request, context):
        # Implement logic to verify email
        return empty_pb2.Empty()

class ChildrenFollowingService(twitter_pb2_grpc.ChildrenFollowingServicer):
    def FollowChild(self, request, context):
        # Implement logic to follow a child
        return empty_pb2.Empty()

    def ListFollowers(self, request, context):
        # Implement logic to list followers of a child
        followers = []
        for follower in followers:
            yield follower

    def ListFollowing(self, request, context):
        # Implement logic to list following of a child
        following = []
        for followed in following:
            yield followed

class TweetsService(twitter_pb2_grpc.TweetsServicer):
    def CreateTweet(self, request, context):
        # Implement logic to create a tweet
        tweet = twitter_pb2.Tweet()
        return tweet

    def GetTweet(self, request, context):
        # Implement logic to get a tweet
        tweet = twitter_pb2.Tweet()
        return tweet

    def DeleteTweet(self, request, context):
        # Implement logic to delete a tweet
        return empty_pb2.Empty()

    def LikeTweet(self, request, context):
        # Implement logic to like a tweet
        return empty_pb2.Empty()

    def ListTweets(self, request, context):
        # Implement logic to list tweets
        tweets = []
        for tweet in tweets:
            yield tweet

    def ListFeedTweets(self, request, context):
        # Implement logic to list feed tweets
        feed_tweets = []
        for tweet in feed_tweets:
            yield tweet

    def CreateTweetComment(self, request, context):
        # Implement logic to create a tweet comment
        tweet_comment = twitter_pb2.TweetComment()
        return tweet_comment

    def GetTweetComment(self, request, context):
        # Implement logic to get a tweet comment
        tweet_comment = twitter_pb2.TweetComment()
        return tweet_comment

def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    twitter_pb2_grpc.add_AccountServicer_to_server(AccountService(), server)
    twitter_pb2_grpc.add_ChildrenFollowingServicer_to_server(ChildrenFollowingService(), server)
    twitter_pb2_grpc.add_TweetsServicer_to_server(TweetsService(), server)
    server.add_insecure_port('[::]:50051')
    server.start()
    server.wait_for_termination()

if __name__ == '__main__':
    serve()




# client side:


# Get a child by username
def get_child_by_username(username):
    request = GetChildByUsernameRequest(username=username)
    try:
        response = client.GetChildByUsername(request)
        print('Child details:', response)
    except grpc.RpcError as e:
        print('Error getting child by username:', e.details())
